istioctl/cmd/generate_cert_key.go

```diff
@@ -0,0 +1,224 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"crypto"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"istio.io/istio/security/pkg/k8s/chiron"
+	"istio.io/istio/security/pkg/pki/util"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"istio.io/pkg/log"
+
+	"istio.io/istio/security/pkg/pki/ca"
+)
+
+var (
+	validFrom string
+	validFor  time.Duration
+	org       string
+	outCert   string
+	outPriv   string
+	keySize   int
+	rootCert  string
+)
+
+const (
+	// caCertID is the CA certificate chain file.
+	caCertID = "ca-cert.pem"
+	// caPrivateKeyID is the private key file of CA.
+	caPrivateKeyID = "ca-key.pem"
+	// CASecret stores the key/cert of self-signed CA for persistency purpose.
+	CASecret = "istio-ca-secret"
+	// CertChainID is the ID/name for the certificate chain file.
+	CertChainID = "cert-chain.pem"
+	// PrivateKeyID is the ID/name for the private key file.
+	PrivateKeyID = "key.pem"
+	// The size of a private key for a self-signed Istio CA.
+	caKeySize  = 2048
+	timeLayout = "Jan 2 15:04:05 2006"
+)
+
+func generateKeyAndCertCmd() *cobra.Command {
+	generateKeyAndCertCmd := &cobra.Command{
+		Use:   "generate-key-cert",
+		Short: "Generate key and cert for workloads",
+		Example: `  # Generate key and certificate for workload workload-name.vmnamespace and extract root certificate.
+  istioctl generate-key-cert  <workload-name.namespace>
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return fmt.Errorf("workload name.namespace is required")
+			}
+			if len(args) > 1 {
+				return fmt.Errorf("only workload-name.vmnamespace is supported")
+			}
+			client, err := interfaceFactory(kubeconfig)
+			if err != nil {
+				return err
+			}
+			writer := cmd.OutOrStdout()
+			return generateCert(args[0], client, writer)
+		},
+	}
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&validFrom, "start-date", "", "Creation date in format of "+timeLayout)
+	generateKeyAndCertCmd.PersistentFlags().DurationVar(&validFor, "duration", 365*24*time.Hour, "Duration that certificate is valid for.")
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&org, "organization", "Juju org", "Organization for the cert.")
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&outCert, "out-cert", "cert-chain.pem", "Output certificate file.")
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&outPriv, "out-priv", "key.pem", "Output private key file.")
+	generateKeyAndCertCmd.PersistentFlags().IntVar(&keySize, "key-size", 2048, "Size of the generated private key")
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&rootCert, "root-cert", "root-cert.pem", "Output root certificate file")
+	return generateKeyAndCertCmd
+}
+
```

Comment:
My inclination here would be to move everything from here down into `istioctl/pkg`. Generally, the `cmd` directory should just be for CLI specific mappings, no functionality.

Reply:
done